sway-lsp/src/capabilities/semantic_tokens.rs

```diff
@@ -1,174 +1,174 @@
-    for item in token_map.iter() {
-        let ((_, span), token) = item.pair();
-        let token_type_idx = type_idx(&token.parsed);
-        let semantic_token = semantic_token(token_type_idx, span, prev_token_span);
+    /// Push a new token onto the builder
+    pub(crate) fn push(&mut self, range: Range, token_index: u32, modifier_bitset: u32) {
+        let mut push_line = range.start.line as u32;
+        let mut push_char = range.start.character as u32;
 
-        semantic_tokens.push(semantic_token);
-        prev_token_span = Some(span.clone());
-    }
+        if !self.data.is_empty() {
+            push_line -= self.prev_line;
+            if push_line == 0 {
+                push_char -= self.prev_char;
+            }
+        }
 
-    semantic_tokens
-}
+        // A token cannot be multiline
+        let token_len = range.end.character - range.start.character;
 
-fn semantic_token(
-    token_type_idx: u32,
-    next_token_span: &Span,
-    prev_token_span: Option<Span>,
-) -> SemanticToken {
-    let next_token_range = get_range_from_span(next_token_span);
-    let next_token_line_start = next_token_range.start.line;
-
-    // TODO - improve with modifiers
-    let token_modifiers_bitset = 0;
-    let length = next_token_range.end.character - next_token_range.start.character;
-
-    let next_token_start_char = next_token_range.start.character;
-
-    let (delta_line, delta_start) = if let Some(prev_token_span) = prev_token_span {
-        let prev_token_range = get_range_from_span(&prev_token_span);
-        let prev_token_line_start = prev_token_range.start.line;
-        let delta_start = if next_token_line_start == prev_token_line_start {
-            next_token_start_char - prev_token_range.start.character
-        } else {
-            next_token_start_char
+        let token = SemanticToken {
+            delta_line: push_line,
+            delta_start: push_char,
+            length: token_len as u32,
+            token_type: token_index,
+            token_modifiers_bitset: modifier_bitset,
         };
-        (next_token_line_start - prev_token_line_start, delta_start)
-    } else {
-        (next_token_line_start, next_token_start_char)
-    };
-
-    SemanticToken {
-        token_modifiers_bitset,
-        token_type: token_type_idx,
-        length,
-        delta_line,
-        delta_start,
+
+        self.data.push(token);
+
+        self.prev_line = range.start.line as u32;
+        self.prev_char = range.start.character as u32;
+    }
+
+    pub(crate) fn build(self) -> SemanticTokens {
+        SemanticTokens {
+            result_id: Some(self.id),
+            data: self.data,
+        }
     }
 }
 
-/// these values should reflect indexes in `token_types`
-#[repr(u32)]
-enum TokenTypeIndex {
-    Function = 1,
-    Parameter = 5,
-    String = 6,
-    Variable = 9,
-    Enum = 10,
-    Struct = 11,
-    Interface = 12,
+static TOKEN_RESULT_COUNTER: AtomicU32 = AtomicU32::new(1);
```

Comment:
Rather than using a global, is it possible to track this in the Session or something along those lines?

That said this looks pretty harmless, and appears to only ever increment upon finishing each token collection. If it's easier to keep it, maybe we could move this static into the `semantic_tokens` function to make it clear that that's the only place it is used?

Reply:
Yeah cool, i'll move it into the `semantic_tokens` function.